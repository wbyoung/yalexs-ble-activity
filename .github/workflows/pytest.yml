# yamllint disable-line rule:line-length
# see: https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-python

name: Pytest

on:
  push:
    branches: [main]
  pull_request:
    branches: [main]
  workflow_dispatch:
    inputs:
      upgraded:
        description: "Include upgraded dependencies in build matrix"
        default: false
        type: boolean
  schedule:
    - cron: "0 0 * * *"

jobs:
  info:
    name: Collect info
    outputs:
      dependencies: ${{ steps.info.outputs.dependencies }}
    runs-on: ubuntu-latest
    steps:
      - name: Collect info
        id: info
        run: |
          dependencies='["pinned", "upgraded"]'

          if [[ "${{ github.event_name }}" == "push" ]] || \
             [[ "${{ github.event_name }}" == "pull_request" ]] || \
             [[ "${{ github.event.inputs.upgraded }}" == "false" ]];
          then
            dependencies='["pinned"]'
          fi

          # Output & sent to GitHub Actions
          echo "dependencies: ${dependencies}"
          echo "dependencies=${dependencies}" >> $GITHUB_OUTPUT
  build:
    name: Run tests
    runs-on: ubuntu-latest
    needs:
      - info
    strategy:
      fail-fast: false
      matrix:
        dependencies: ${{ fromJSON(needs.info.outputs.dependencies) }}
        python-version: ["3.13"]

    steps:
      - uses: actions/checkout@v4
      - name: Set up Python ${{ matrix.python-version }}
        uses: actions/setup-python@v5
        with:
          python-version: ${{ matrix.python-version }}
      - name: Install dependencies
        run: |
          python -m pip install --upgrade pip

          if [[ "${{ matrix.dependencies }}" == "upgraded" ]];
          then
            echo "unpinning dependencies"
            ./script/unfreeze
          fi

          pip install -r requirements.test.txt

          if [[ "${{ matrix.dependencies }}" == "upgraded" ]];
          then
            ./script/freeze
            echo "dependency changes after upgrade:"
            git diff requirements.test.txt
          fi
      - name: Test with pytest
        run: |
          pytest -p no:sugar
      - name: Validate coverage
        run: |
          coverage json -q -o - | python3 .github/validate_coverage.py
