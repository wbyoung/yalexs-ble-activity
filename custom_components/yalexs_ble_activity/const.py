"""Constants for the Yale Access Bluetooth Activity integration."""

from typing import Final

DOMAIN: Final = "yalexs_ble_activity"
YALEXSBLE_PATCH_URL = (
    "git+https://github.com/wbyoung/yalexs-ble@yalexs-ble-{version}-patches"
)

ATTR_REMOTE_TYPE: Final = "remote_type"
ATTR_SLOT: Final = "slot"
ATTR_SOURCE: Final = "source"
ATTR_TIMESTAMP: Final = "timestamp"

CONF_LOCK_ENTITIES: Final = "lock_entities"

OPERATION_SENSOR_WRITE_DELAY: Final = 2

TRACE: Final = 5
