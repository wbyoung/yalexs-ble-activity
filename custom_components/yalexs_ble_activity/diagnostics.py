"""Diagnostics support for Yale Access Bluetooth Activity."""

from typing import Any

from homeassistant.components.diagnostics import async_redact_data
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant

TO_REDACT: set[str] = set()


async def async_get_config_entry_diagnostics(  # noqa: RUF029
    hass: HomeAssistant,  # noqa: ARG001
    entry: ConfigEntry,
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    result: dict[str, Any] = async_redact_data(entry.as_dict(), TO_REDACT)
    return result
