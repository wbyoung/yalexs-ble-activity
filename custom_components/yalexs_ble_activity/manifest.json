{
  "domain": "yalexs_ble_activity",
  "name": "Yale Access Bluetooth Activity",
  "after_dependencies": ["recorder"],
  "codeowners": [
    "@wbyoung"
  ],
  "config_flow": true,
  "dependencies": ["yalexs_ble"],
  "documentation": "https://github.com/wbyoung/yalexs-ble-activity",
  "homekit": {},
  "iot_class": "calculated",
  "issue_tracker": "https://github.com/wbyoung/yalexs-ble-activity/issues",
  "requirements": [],
  "single_config_entry": true,
  "ssdp": [],
  "version": "0.1.1",
  "zeroconf": []
}
