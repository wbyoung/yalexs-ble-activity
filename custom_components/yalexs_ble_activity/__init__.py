"""The Yale Access Bluetooth Activity integration."""

from __future__ import annotations

from functools import partial
from importlib.metadata import version
import logging

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import Platform
from homeassistant.core import Event, HomeAssistant
from homeassistant.exceptions import ConfigEntryError
from homeassistant.helpers import config_validation as cv, entity_registry as er
from homeassistant.helpers.event import async_track_entity_registry_updated_event
from homeassistant.helpers.issue_registry import IssueSeverity, async_create_issue
from homeassistant.util import package as pkg_util
import yalexs_ble

from .const import CONF_LOCK_ENTITIES, DOMAIN, YALEXSBLE_PATCH_URL

_LOGGER = logging.getLogger(__name__)

PLATFORMS: list[Platform] = [Platform.SENSOR]
CONFIG_SCHEMA = cv.config_entry_only_config_schema(DOMAIN)
YALEXSBLE_VERSION = version("yalexs-ble")


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up Yale Access Bluetooth Activity from a config entry.

    Returns:
        If the setup was successful.

    Raises:
        ConfigEntryError: If there is an issue with setup.
    """
    _LOGGER.debug("setup %s with config:%s", entry.title, entry.data)

    if not hasattr(yalexs_ble.PushLock, "register_activity_callback"):
        _LOGGER.debug("%s no activity callback:%s", entry.title, entry.data)

        def _install_yalexs_ble_patch() -> bool:
            return bool(
                pkg_util.install_package(
                    YALEXSBLE_PATCH_URL.format(version=YALEXSBLE_VERSION)
                )
            )

        raise ConfigEntryError(
            translation_domain=DOMAIN,
            translation_key="yalexs_ble_patched"
            if (await hass.async_add_executor_job(_install_yalexs_ble_patch))
            else "yalexs_ble_no_patch_available",
            translation_placeholders={
                "yalexs_ble_version": YALEXSBLE_VERSION,
            },
        )

    entry.async_on_unload(
        async_track_entity_registry_updated_event(
            hass,
            entry.data[CONF_LOCK_ENTITIES],
            partial(_async_handle_lock_entity_change, hass, entry),
        ),
    )

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
    return True


async def _async_handle_lock_entity_change(  # noqa: RUF029
    hass: HomeAssistant,
    entry: ConfigEntry,
    event: Event[er.EventEntityRegistryUpdatedData],
) -> None:
    """Fetch and process tracked entity change event."""
    data = event.data
    lock_entity_ids: list[str] = entry.data[CONF_LOCK_ENTITIES]

    if data["action"] == "remove":
        _create_removed_lock_entity_issue(hass, data["entity_id"])

    if data["action"] == "update" and "entity_id" in data["changes"]:
        old_lock_id = data["old_entity_id"]
        new_lock_id = data["entity_id"]

        hass.config_entries.async_update_entry(
            entry,
            data={
                **entry.data,
                CONF_LOCK_ENTITIES: [
                    new_lock_id if lock_id == old_lock_id else lock_id
                    for lock_id in lock_entity_ids
                ],
            },
        )


def _create_removed_lock_entity_issue(
    hass: HomeAssistant,
    entity_id: str,
) -> None:
    """Create a repair issue for a removed lock entity."""
    async_create_issue(
        hass,
        DOMAIN,
        f"lock_entity_removed_{entity_id}",
        is_fixable=True,
        is_persistent=True,
        severity=IssueSeverity.WARNING,
        translation_key="lock_entity_removed",
        translation_placeholders={
            "entity_id": entity_id,
        },
    )
