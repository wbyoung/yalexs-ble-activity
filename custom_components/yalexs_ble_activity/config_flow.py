"""Config flow for Yale Access Bluetooth Activity integration."""

from __future__ import annotations

from typing import Any

from homeassistant.components.lock import DOMAIN as LOCK_DOMAIN
from homeassistant.components.yalexs_ble.const import DOMAIN as YALEXSBLE_DOMAIN
from homeassistant.config_entries import (
    ConfigEntry,
    ConfigFlow,
    ConfigFlowResult,
    OptionsFlow,
)
from homeassistant.core import callback
from homeassistant.helpers.selector import EntitySelector, EntitySelectorConfig
import voluptuous as vol

from .const import CONF_LOCK_ENTITIES, DOMAIN

STEP_USER_DATA_SCHEMA = vol.Schema(
    {
        vol.Required(
            CONF_LOCK_ENTITIES,
        ): EntitySelector(
            EntitySelectorConfig(
                integration=YALEXSBLE_DOMAIN,
                domain=[LOCK_DOMAIN],
                multiple=True,
            ),
        ),
    }
)


class YaleXSBLEActivityConfigFlow(ConfigFlow, domain=DOMAIN):  # type: ignore[call-arg]
    """Handle a Yale Access Bluetooth Activity config flow."""

    VERSION = 1

    @staticmethod
    @callback
    def async_get_options_flow(
        config_entry: ConfigEntry,  # noqa: ARG004
    ) -> OptionsFlow:
        """Get the options flow for this handler.

        Returns:
            The options flow.
        """
        return YaleXSBLEActivityOptionsFlow()

    async def async_step_user(
        self,
        user_input: dict[str, Any] | None = None,
    ) -> ConfigFlowResult:
        """Handle a flow initialized by the user.

        Returns:
            The config flow result.
        """
        if user_input is not None:
            return self.async_create_entry(
                title="Yale Access Bluetooth Activity", data=user_input
            )

        return self.async_show_form(
            step_id="user",
            data_schema=self.add_suggested_values_to_schema(
                STEP_USER_DATA_SCHEMA, user_input
            ),
        )


class YaleXSBLEActivityOptionsFlow(OptionsFlow):
    """Handle a option flow."""

    async def async_step_init(
        self,
        user_input: dict[str, Any] | None = None,
    ) -> ConfigFlowResult:
        """Handle options flow.

        Returns:
            The config flow result.
        """
        if user_input is not None:
            self.hass.config_entries.async_update_entry(
                self.config_entry,
                data={**self.config_entry.data, **user_input},
            )
            return self.async_create_entry(data={})

        return self.async_show_form(
            step_id="init",
            data_schema=self.add_suggested_values_to_schema(
                STEP_USER_DATA_SCHEMA,
                self.config_entry.data if user_input is None else user_input,
            ),
        )
