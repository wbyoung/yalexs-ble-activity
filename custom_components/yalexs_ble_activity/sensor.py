"""Support for Yale Access Bluetooth Activity sensors."""

from __future__ import annotations

import datetime as dt
import logging
from typing import Any

from homeassistant.components import recorder
from homeassistant.components.sensor import SensorEntity
from homeassistant.components.yalexs_ble.entity import YALEXSBLEEntity
from homeassistant.components.yalexs_ble.models import YaleXSBLEData
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import EVENT_STATE_CHANGED, STATE_UNAVAILABLE, STATE_UNKNOWN
from homeassistant.core import (
    CALLBACK_TYPE,
    Event,
    EventStateChangedData,
    HomeAssistant,
    State,
    callback,
)
from homeassistant.helpers import entity_registry as er, event as evt
from homeassistant.helpers.entity_platform import AddConfigEntryEntitiesCallback
from homeassistant.helpers.restore_state import (
    ExtraStoredData,
    RestoredExtraData,
    RestoreEntity,
)
from homeassistant.util import dt as dt_util
from yalexs_ble import ConnectionInfo, DoorActivity, LockActivity, LockInfo

from .const import (
    ATTR_REMOTE_TYPE,
    ATTR_SLOT,
    ATTR_SOURCE,
    ATTR_TIMESTAMP,
    CONF_LOCK_ENTITIES,
    OPERATION_SENSOR_WRITE_DELAY,
)

_LOGGER = logging.getLogger(__name__)


async def async_setup_entry(  # noqa: RUF029
    hass: HomeAssistant,
    entry: ConfigEntry,
    async_add_entities: AddConfigEntryEntitiesCallback,
) -> None:
    """Set up Yale Access Bluetooth Activity sensors."""

    entity_registry = er.async_get(hass)

    async_add_entities(
        YaleXSBLEOperationSensor(data)
        for lock_enitity_id in entry.data[CONF_LOCK_ENTITIES]
        if (
            (lock_entry := entity_registry.async_get(lock_enitity_id))
            and (core_entry_id := lock_entry.config_entry_id)
            and (core_entry := hass.config_entries.async_get_known_entry(core_entry_id))
            and (data := core_entry.runtime_data)
        )
    )


class YaleXSBLEOperationSensor(YALEXSBLEEntity, SensorEntity, RestoreEntity):
    """Representation of an Yale Access Bluetooth lock operation sensor."""

    _attr_translation_key = "operation"
    _attr_icon = "mdi:lock-clock"
    _pending_activity_update: DoorActivity | LockActivity | None = None
    _cancel_pending_activity_update: CALLBACK_TYPE | None = None

    def __init__(
        self,
        data: YaleXSBLEData,
    ) -> None:
        """Initialize the sensor."""
        super().__init__(data)
        self._attr_unique_id = f"{data.lock.address}operation"

    @callback
    def _async_activity_update(
        self,
        activity: DoorActivity | LockActivity,
        lock_info: LockInfo,  # noqa: ARG002
        connection_info: ConnectionInfo,  # noqa: ARG002
    ) -> None:
        """Handle activity update."""

        value, attributes = self._extract_values(activity)

        _LOGGER.debug("creating event for activity update")

        self.hass.bus.async_fire(
            "yalexs_ble_activity",
            {
                "entity_id": self.entity_id,
                "state": value,
                "attributes": attributes,
            },
        )

        if self._pending_activity_update:
            self._record_pending_update()

        self._pending_activity_update = activity

        if self._cancel_pending_activity_update:
            self._cancel_pending_activity_update()

        self._cancel_pending_activity_update = evt.async_call_later(
            self.hass,
            OPERATION_SENSOR_WRITE_DELAY,
            self._flush_pending_update,
        )

    def _record_pending_update(self) -> None:
        activity = self._pending_activity_update
        assert activity is not None
        native_value, attributes = self._extract_values(activity)
        state_changed_data: EventStateChangedData = {
            "entity_id": self.entity_id,
            "old_state": None,
            "new_state": State(
                self.entity_id,
                native_value or STATE_UNAVAILABLE,
                attributes,
                last_changed=activity.timestamp,
                last_reported=activity.timestamp,
                last_updated=activity.timestamp,
                last_updated_timestamp=dt_util.as_timestamp(activity.timestamp),
            ),
        }

        _LOGGER.debug("writing historic activity update: %s", state_changed_data)

        instance = recorder.get_instance(self.hass)
        instance.queue_task(Event(str(EVENT_STATE_CHANGED), state_changed_data))

    @callback
    def _flush_pending_update(self, now: dt.datetime) -> None:  # noqa: ARG002
        activity = self._pending_activity_update
        assert activity is not None

        _LOGGER.debug("flushing pending activity update")

        self._attr_native_value, self._attr_extra_state_attributes = (
            self._extract_values(activity)
        )
        self._pending_activity_update = None

        self.async_write_ha_state()

    @staticmethod
    def _extract_values(
        activity: DoorActivity | LockActivity,
    ) -> tuple[str | None, dict[str, Any]]:
        value: str | None = None
        attributes: dict[str, Any] = {}

        if isinstance(activity, DoorActivity):
            value = f"door_{activity.status.name.lower()}"
            attributes[ATTR_TIMESTAMP] = activity.timestamp
        elif isinstance(activity, LockActivity):
            value = f"lock_{activity.status.name.lower()}"
            attributes[ATTR_TIMESTAMP] = activity.timestamp
            attributes[ATTR_SOURCE] = activity.source.name.lower()
            if activity.remote_type is not None:
                attributes[ATTR_REMOTE_TYPE] = activity.remote_type.name.lower()
            if activity.slot is not None:
                attributes[ATTR_SLOT] = activity.slot

        return (value, attributes)

    async def async_added_to_hass(self) -> None:
        """Register callbacks, perform initial updates & restore state."""
        await super().async_added_to_hass()

        self.async_on_remove(
            self._device.register_activity_callback(
                self._async_activity_update, request_update=True
            )
        )

        if (
            (last_state := await self.async_get_last_state()) is not None
            and last_state.state not in {STATE_UNKNOWN, STATE_UNAVAILABLE}
            and (extra_data := await self.async_get_last_extra_data()) is not None
        ):
            extra_data_dict = extra_data.as_dict()
            self._attr_native_value = extra_data_dict["value"]
            self._attr_extra_state_attributes = extra_data_dict["attributes"]

    @property
    def extra_restore_state_data(self) -> ExtraStoredData | None:
        return RestoredExtraData(
            {
                "value": self._attr_native_value,
                "attributes": self._attr_extra_state_attributes,
            }
        )
