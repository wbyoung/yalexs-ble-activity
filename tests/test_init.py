"""Test component setup."""

from unittest.mock import patch

from homeassistant.config_entries import ConfigEntryState
from homeassistant.core import HomeAssistant
from homeassistant.helpers import (
    device_registry as dr,
    entity_registry as er,
    issue_registry as ir,
)
from homeassistant.setup import async_setup_component
import pytest
from pytest_homeassistant_custom_component.common import MockConfigEntry
from syrupy.assertion import SnapshotAssertion
from syrupy.filters import props

from custom_components.yalexs_ble_activity import YALEXSBLE_VERSION
from custom_components.yalexs_ble_activity.const import CONF_LOCK_ENTITIES, DOMAIN

from . import setup_integration


async def test_async_setup(hass: HomeAssistant):
    """Test the component gets setup."""
    assert await async_setup_component(hass, DOMAIN, {}) is True


async def test_standard_setup(
    hass: HomeAssistant,
    config_entry: MockConfigEntry,
    lock: er.RegistryEntry,
    device_registry: dr.DeviceRegistry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test standard setup."""
    await setup_integration(hass, config_entry)

    device = device_registry.async_get(lock.device_id)

    assert device is not None
    assert device.id == lock.device_id
    assert device == snapshot(
        name="device",
        exclude=props(
            # compat for HA DeviceRegistryEntrySnapshot <2025.8.0 and >=2025.6.1
            "suggested_area",
            # compat for HA DeviceRegistryEntrySnapshot <2025.9.0 and >=2025.6.1
            "is_new",
        ),
    )


@pytest.mark.parametrize(
    "scenario",
    [
        (True, "Restart required to use newly patched `yalexs_ble` package"),
        (
            False,
            f"No patches for `yalexs_ble=={YALEXSBLE_VERSION}`; one must be created",
        ),
    ],
    ids=["successful_install", "failed_install"],
)
async def test_patching_yalexs_ble(
    hass: HomeAssistant,
    config_entry: MockConfigEntry,
    scenario: tuple[bool, str],
    lock: er.RegistryEntry,
    device_registry: dr.DeviceRegistry,
    snapshot: SnapshotAssertion,
    caplog: pytest.LogCaptureFixture,
) -> None:
    """Test standard setup."""

    install_success, expected_error = scenario

    with (
        patch("yalexs_ble.PushLock") as push_lock_class_mock,
        patch(
            "homeassistant.util.package.install_package", return_value=install_success
        ),
    ):
        del push_lock_class_mock.register_activity_callback

        await setup_integration(hass, config_entry)
        assert config_entry.state is ConfigEntryState.SETUP_ERROR
        assert expected_error in caplog.text


async def test_renamed_lock_entity(
    hass: HomeAssistant,
    config_entry: MockConfigEntry,
    lock: er.RegistryEntry,
    device_registry: dr.DeviceRegistry,
    entity_registry: er.EntityRegistry,
) -> None:
    """Test that config entry data is updated when lock entity is renamed."""
    hass.states.async_set(lock.entity_id, "locked")

    assert config_entry.data[CONF_LOCK_ENTITIES] == [lock.entity_id]
    await setup_integration(hass, config_entry)

    entity_registry.async_update_entity(
        lock.entity_id,
        new_entity_id=f"{lock.entity_id}_renamed",
    )
    await hass.async_block_till_done()
    assert config_entry.data[CONF_LOCK_ENTITIES] == [f"{lock.entity_id}_renamed"]


async def test_create_removed_lock_entity_issue(
    hass: HomeAssistant,
    config_entry: MockConfigEntry,
    lock: er.RegistryEntry,
    issue_registry: ir.IssueRegistry,
    entity_registry: er.EntityRegistry,
) -> None:
    """Test we create an issue for removed lock entities."""
    hass.states.async_set(lock.entity_id, "locked")

    await setup_integration(hass, config_entry)

    hass.states.async_remove(lock.entity_id)
    entity_registry.async_remove(lock.entity_id)
    await hass.async_block_till_done()

    assert issue_registry.async_get_issue(
        DOMAIN,
        f"lock_entity_removed_{lock.entity_id}",
    )
