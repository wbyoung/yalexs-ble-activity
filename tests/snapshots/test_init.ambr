# serializer version: 1
# name: test_standard_setup[device]
  DeviceRegistryEntrySnapshot({
    'area_id': None,
    'config_entries': <ANY>,
    'config_entries_subentries': <ANY>,
    'configuration_url': None,
    'connections': set({
      tuple(
        'bluetooth',
        'mock-address:front_door',
      ),
    }),
    'disabled_by': None,
    'entry_type': None,
    'hw_version': None,
    'id': <ANY>,
    'identifiers': set({
      tuple(
        'yalexs_ble',
        'mock-serial:front_door',
      ),
    }),
    'labels': set({
    }),
    'manufacturer': 'mock-manufacturer:front_door',
    'model': 'mock-model:front_door',
    'model_id': None,
    'name': 'Front door',
    'name_by_user': None,
    'primary_config_entry': <ANY>,
    'serial_number': None,
    'sw_version': 'mock-firmware:front_door',
    'via_device_id': None,
  })
# ---
